### ISSUE TITLE
[aws/ebs] nvm support

### ISSUE THREAD

**Issue by jippi** [thecodeteam/rexray#1104]
# Summary

When using EBS on a `c5.*` instance (using nvm) the volume are mounted under a different name than non-nvm instances.

# Bug Reports

[c5](https://aws.amazon.com/blogs/aws/now-available-compute-intensive-c5-instances-for-amazon-ec2/) and other nvm instances mount volumes under a different `/dev` path than non-nvm instances. 

The path is `/dev/nvme*n*` instead of `/dev/sd*` or `/dev/xvd*` - apparently NVMe ignores the volume mount completely and always assign them in a `${prev}+1` naming scheme

I'm not in any way expert in EC2 / NVMe or how these things could work, I'm just putting it out here than our rexray usage as docker plugin did not work on c5 due to the path difference

## Version

Any version

**Comment by clintkitson** [comment]
Very interesting @jippi.

Thank you for the writeup and links for reference for implementation. From what I can see so far this involves 

0) discovery of valid toolset for introspecting device and volumeid relationships
1) update of ebs container to include such tooling, paths, and testing across ami/typical envs
2) new awareness of whether devices are showing as nvm or typical ebs
3) new discovery steps for nvm devices
4) skipping device picking logic
5) pre-detach command to ensure flush since a forced detach is always requested with nvme

**Comment by clintkitson** [comment]
Agree, that's why I called it out as an update that is needed for the EBS driver =)

**Comment by djannot** [comment]
+1

**Comment by clintkitson** [comment]
We are currently working through the plans for the upcoming REX-Ray releases inclusive of a path forward to native CSI drivers. We wanted to combine this phase while including enhanced and additional functionality that goes beyond what currently exists in REX or the K8s ebs plugin.

**Comment by matthewmrichter** [comment]
+1

**Comment by jippi** [comment]
Whats the status on this @clintkitson ? :)

**Comment by clintkitson** [comment]
Sounds amazing =)

Pinging @arun-gupta re AWS and CSI where this functionality would make the most sense.

**Comment by arun-gupta** [comment]
I've asked @aaithal to look into this

**Comment by aaithal** [comment]
@clintkitson  your [comment](https://github.com/rexray/rexray/issues/1104#issuecomment-358689102) makes sense to me. The driver being responsible for discovery and subsequent pass-through sounds good as well.

**Comment by jippi** [comment]
So timeline wise, which quarter of 2018 might 5th gen AWS support land?

**Comment by clintkitson** [comment]
Unfortunately there is no commitment to this functionality as of yet. Looking forward to working with AWS's team around a CSI driver for EBS to bring this forward. @aaithal

**Comment by jippi** [comment]
@clintkitson thats a frustrating answer, its basically blocking all rexray users to look elsewhere for volume management on AWS or stick to previous-generation instances :( Any way to help guide the prioritisation on this to be higher?

**Comment by dispalt** [comment]
does that mean this is completely dependent on a new implementation around CSI before it would be fixed on rexray, or is that still up in the air?

**Comment by clintkitson** [comment]
Yes for the project we are planning on the support in CSI. We'd be happy to look at a contribution that has this functioning in the meantime.

**Comment by jippi** [comment]
@clintkitson would it be acceptable to just check if `/dev/<device name>` exist for nvme?

**Comment by JustinVenus** [comment]
I have a dirty work around to support EBS on M5/C5's.  It requires some code changes, udev rules, and nvme-cli.  Here is the gist of what I did to make this work for me https://gist.github.com/JustinVenus/7d26ff5e885dc0ed34951ce20e5e79e0 ... I could make a proper pull request if it is wanted.

**Comment by jippi** [comment]
@JustinVenus looks good! can you submit the changes as a PR ? :)

**Comment by JustinVenus** [comment]
I opened a PR for this issue at https://github.com/rexray/rexray/pull/1233.

**Comment by nunofernandes** [comment]
This (https://github.com/seatgeek/rexray/commit/c01de505dfefdc1ea8cc4eb74f73ce22dc2935ed) worked for me...

**Comment by gai00** [comment]
waiting for docker plugin update for this.

**Comment by raravena80** [comment]
👍

**Comment by jamesdh** [comment]
Just an FYI that this applies to all C5, C5d, i3.metal, M5, M5d, R5, R5d, T3, and z1d instance types. 

T3 being the new lower-end general purpose tier. This issue pretty much kills rexray for any modern EC2 instance.

**Comment by clintkitson** [comment]
Thanks for the clarification @jamesdh

**Comment by CameronGo** [comment]
Does this PR close this issue out? https://github.com/rexray/rexray/pull/1252

**Comment by jippi** [comment]
A new release have to be cut, but otherwise, yes

**Comment by jippi** [comment]
@akutz @clintkitson when do you plan to cut a release? :)

**Comment by clintkitson** [comment]
@jippi Did you get a chance to test the latest unstable release?

**Comment by dovreshef** [comment]
Hi

I'm also running into this. Any plan to release soon?

**Comment by digovc** [comment]
I'm using the master version on T3 instances with the PR from @jippi and really works.

Please release a new version with master to resolve this problem with the new EBS NVMe blocks.

**Comment by clintkitson** [comment]
If we can verify the `unstable` released binaries work as expected for AWS then we can proceed with a release. Can someone signup to verify that? Thank you!

**Comment by matthewmrichter** [comment]
Isn't that what @digovc said he did?

**Comment by clintkitson** [comment]
Apologies for not being more clear. Below is the link for unstable binaries that were built from the merge of this fix. We are looking for testers on this build.

https://dl.bintray.com/rexray/rexray/unstable/0.11.3+36/

**Comment by dovreshef** [comment]
@clintkitson 
I don't think that version includes the fix. I have installed that version with:
```bash
curl -sSL https://rexray.io/install | sh -s -- unstable 0.11.3+36
```

I still get the same error with it, and in the rexray service logs I see:
```
Jan 08 09:25:17 ip-10-200-2-147.eu-central-1.compute.internal rexray[3566]: time="2019-01-08T09:25:17Z" level=warning msg="device does not match" deviceName=nvme0n1 deviceRX=^xvd[f-p]$ host="unix:///var/run/rexray/015457837.sock" instanceID="ebs=i-0252da03a1b8be6c7,availabilityZone=eu-central-1c&region=eu-central-1" integrationDriver=linux osDriver=linux server=cedar-face-ly service=ebs storageDriver=libstorage time=1546939517619 txCR=1546939517 txID=3e34070a-af62-4dbd-4b3b-811463a4e039
```

**Comment by dovreshef** [comment]
@clintkitson 
I don't think that version includes the fix. I have installed that version with:
```bash
curl -sSL https://rexray.io/install | sh -s -- unstable 0.11.3+36
```

I still get the same error with it, and in the rexray service logs I see:
```
Jan 08 09:25:17 ip-10-200-2-147.eu-central-1.compute.internal rexray[3566]: time="2019-01-08T09:25:17Z" level=warning msg="device does not match" deviceName=nvme0n1 deviceRX=^xvd[f-p]$ host="unix:///var/run/rexray/015457837.sock" instanceID="ebs=i-XXXXXX,availabilityZone=eu-central-1c&region=eu-central-1" integrationDriver=linux osDriver=linux server=cedar-face-ly service=ebs storageDriver=libstorage time=1546939517619 txCR=1546939517 txID=3e34070a-af62-4dbd-4b3b-811463a4e039
```

**Comment by clintkitson** [comment]
@dovreshef Can you make a suggestion for the documentation updates or better yet a PR including them? Thanks for the feedback.

**Comment by edthorne** [comment]
I think there's a subtle issue that's lingering here. I'm seeing the following output by the docker daemon frequently:

```
Feb 11 19:37:16 ip-172-31-53-17 dockerd: time="2019-02-11T19:37:16Z" level=error msg="time=\"2019-02-11T19:37:16Z\" level=warning msg=\"device does not match\" deviceName=name deviceRX=^xvd[f-p]$ host=\"unix:///var/run/rexray/239479381.sock\" instanceID=\"ebs=i-XXXXXXXXXXXXXXXXX,availabilityZone=us-east-1a&region=us-east-1\" integrationDriver=linux osDriver=linux server=bevel-minnow-ie service=ebs storageDriver=libstorage time=1549913836132 txCR=1549913836 txID=9bf36cdd-8437-4ec8-7178-7ab401ea062f " plugin=0e1b330cbb644b81c26a9cce14c76ede21f347829c53d53d53832027e69fcc53
Feb 11 19:37:16 ip-172-31-53-17 dockerd: time="2019-02-11T19:37:16Z" level=error msg="time=\"2019-02-11T19:37:16Z\" level=warning msg=\"device does not match\" deviceName=nvme1n1 deviceRX=^xvd[f-p]$ host=\"unix:///var/run/rexray/239479381.sock\" instanceID=\"ebs=i-XXXXXXXXXXXXXXXXX,availabilityZone=us-east-1a&region=us-east-1\" integrationDriver=linux osDriver=linux server=bevel-minnow-ie service=ebs storageDriver=libstorage time=1549913836133 txCR=1549913836 txID=9bf36cdd-8437-4ec8-7178-7ab401ea062f " plugin=0e1b330cbb644b81c26a9cce14c76ede21f347829c53d53d53832027e69fcc53
Feb 11 19:37:16 ip-172-31-53-17 dockerd: time="2019-02-11T19:37:16Z" level=error msg="time=\"2019-02-11T19:37:16Z\" level=warning msg=\"device does not match\" deviceName=nvme0n1 deviceRX=^xvd[f-p]$ host=\"unix:///var/run/rexray/239479381.sock\" instanceID=\"ebs=i-XXXXXXXXXXXXXXXXX,availabilityZone=us-east-1a&region=us-east-1\" integrationDriver=linux osDriver=linux server=bevel-minnow-ie service=ebs storageDriver=libstorage time=1549913836133 txCR=1549913836 txID=9bf36cdd-8437-4ec8-7178-7ab401ea062f " plugin=0e1b330cbb644b81c26a9cce14c76ede21f347829c53d53d53832027e69fcc53
Feb 11 19:37:16 ip-172-31-53-17 dockerd: time="2019-02-11T19:37:16Z" level=error msg="time=\"2019-02-11T19:37:16Z\" level=warning msg=\"device does not match\" deviceName=nvme0n1p1 deviceRX=^xvd[f-p]$ host=\"unix:///var/run/rexray/239479381.sock\" instanceID=\"ebs=i-XXXXXXXXXXXXXXXXX,availabilityZone=us-east-1a&region=us-east-1\" integrationDriver=linux osDriver=linux server=bevel-minnow-ie service=ebs storageDriver=libstorage time=1549913836134 txCR=1549913836 txID=9bf36cdd-8437-4ec8-7178-7ab401ea062f " plugin=0e1b330cbb644b81c26a9cce14c76ede21f347829c53d53d53832027e69fcc53
```
Docker believes them to be errors even though rexray appears to report them as warnings.

The mounting of EBS volumes appears to be happening normally and things are working. Here's the relevant version info for our swarms.

```
"DockerVersion": "18.06.1-ce"
"Name": "rexray/ebs:latest"
"PluginReference": "docker.io/rexray/ebs:0.11.4"
 ```

**Comment by gmembre-zenika** [comment]
+1 same problem here : still seeing the error and cannot mount EBS volume :'(

**Comment by gmembre-zenika** [comment]
I mis-read the logs : the driver complains about root devices of the VM but mounting volumes with the plugin works like a charm. Thanks !

**Comment by darkl0rd** [comment]
This is still not working for me.

I have just installed **rexray/ebs:0.11.4** on a t3.2xlarge:

```
$ docker plugin ls
b71744536bc8        rexray/ebs:0.11.4   REX-Ray for Amazon EBS   true
```

When I try to run a container using an EBS volume:
```
$ docker run --tty \
             --interactive \
             --rm \
             --volume-driver rexray/ebs:0.11.4 \
             --volume my_test_volume:/test \
             alpine /bin/sh

docker: Error response from daemon: error while mounting volume '': VolumeDriver.Mount: docker-legacy: Mount: my_test_volume: failed: no device name returned.
```

To (re-)confirm that this is indeed 0.11.4 that is running:
```
$ cat /var/lib/docker/plugins/b71744536bc8704c6b89cb70e9c13fb534616593df4dbec604c91a14ab098c22/rootfs/var/log/rexray/rexray.log

REX-Ray
-------
Binary: /usr/bin/rexray
Flavor: client+agent+controller
SemVer: 0.11.4
OsArch: Linux-x86_64
Commit: e7414eaa971b27977d2283f2882825393493179d
Formed: Tue, 15 Jan 2019 16:03:57 UTC

```

**Comment by MonsieurPaulLeBoulanger** [comment]
@darkl0rd : did you add the required udev rule & associated handler to make it work ?
It's well hidden in rexray documentation :
https://rexray.readthedocs.io/en/v0.11.4/user-guide/storage-providers/aws/#nvme-support

You may find more informations about this issue here:  https://github.com/rexray/rexray/pull/1252 (thanks @jippi )

**Comment by darkl0rd** [comment]
@MonsieurPaulLeBoulanger In all honesty, I did read that - however, I figured using the docker volume plugin that would be unnecessary?

**Comment by dcelasun** [comment]
Having the same problem here with a t3.medium instance. Amazon's own AMI already has a device renaming udev rule, so the one from the rexray docs shouldn't be necessary.

**Comment by paddie** [comment]
Confirmed, but it looks different from the one that @MonsieurPaulLeBoulanger linked to (specifically the `PROGRAM` part) - anything that would break anything?


```bash
$ cat /etc/udev/rules.d/70-ec2-nvme-devices.rules 
# Copyright (C) 2006-2016 Amazon.com, Inc. or its affiliates.
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
#
#    http://aws.amazon.com/apache2.0/
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
# OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the
# License.

#nvme-ns-* devices
KERNEL=="nvme[0-9]*n[0-9]*", ENV{DEVTYPE}=="disk", ATTRS{serial}=="?*", ATTRS{model}=="?*", SYMLINK+="disk/by-id/nvme-$attr{model}_$attr{serial}-ns-%n", OPTIONS+="string_escape=replace"

#nvme partitions
KERNEL=="nvme[0-9]*n[0-9]*p[0-9]*", ENV{DEVTYPE}=="partition", ATTRS{serial}=="?*", ATTRS{model}=="?*",  IMPORT{program}="ec2nvme-nsid %k"
KERNEL=="nvme[0-9]*n[0-9]*p[0-9]*", ENV{DEVTYPE}=="partition", ATTRS{serial}=="?*", ATTRS{model}=="?*",  ENV{_NS_ID}=="?*", SYMLINK+="disk/by-id/nvme-$attr{model}_$attr{serial}-ns-$env{_NS_ID}-part%n", OPTIONS+="string_escape=replace"

# ebs nvme devices
KERNEL=="nvme[0-9]*n[0-9]*", ENV{DEVTYPE}=="disk", ATTRS{model}=="Amazon Elastic Block Store", PROGRAM="/sbin/ebsnvme-id -u /dev/%k", SYMLINK+="%c"
```

**Comment by sp0f** [comment]
Same problem here with version v0.11.4 (on all nvme instance types):
`level=warning msg="device does not match" deviceName=nvme0n1p128 deviceRX=^xvd[f-p]$ host="unix:///var/run/rexray/659941617.sock" (...)`

`level=error msg="docker-legacy: Mount: my-test-vol: failed: no device name returned" host="unix:///var/run/rexray/659941617.sock" integrationDriver=linux osDriver=linux service=ebs storageDriver=libstorage time=1559659515851`

**Comment by RoyHP** [comment]
Same issue here.

**Comment by naviat** [comment]
the latest release (0.11.4) has not solved this issue?

**Comment by jquijino** [comment]
I can also confirm that t3 instances are working with v0.11.4. When adding the udev rule adjustments to those machines, don't for to install the `nvme-cli` package and reload the rules using

```
udevadm control --reload-rules && udevadm trigger
```

**Comment by timhaak** [comment]
For anyone else having problems. Mine turned out to be me messing up creating the files by script.

So here are the scripts that I'm using that eventually worked on t3.xlarge2 (Ubuntu 18.04)

Thanks to the previous comments which helped me find the problem :)

### Make sure nvme is installed

```
apt install nvme-cli -y
```

### Add/etc/udev/rules.d/999-aws-ebs-nvme.rules
```
echo '# /etc/udev/rules.d/999-aws-ebs-nvme.rules
# ebs nvme devices
KERNEL=="nvme[0-9]*n[0-9]*", ENV{DEVTYPE}=="disk", ATTRS{model}=="Amazon Elastic Block Store", PROGRAM="/usr/local/bin/ebs-nvme-mapping /dev/%k", SYMLINK+="%c"' > /etc/udev/rules.d/999-aws-ebs-nvme.rules
```

### Add /usr/local/bin/ebs-nvme-mapping
```
echo '#!/bin/bash
# /usr/local/bin/ebs-nvme-mapping
vol=$(/usr/sbin/nvme id-ctrl --raw-binary "$1" | cut -c3073-3104 | tr -s "'" "'" | sed "'"s/ $//g"'")
vol=${vol#/dev/}
if [[ -n "$vol" ]]; then
  echo ${vol/xvd/sd} ${vol/sd/xvd}
fi' > /usr/local/bin/ebs-nvme-mapping
chmod u+x /usr/local/bin/ebs-nvme-mapping
```
### Make sure rules have run or reboot
udevadm control --reload-rules && udevadm trigger

### Using rexray/ebs:edge
```
docker plugin install --grant-all-permissions rexray/ebs:edge \
  EBS_ACCESSKEY="${EBS_ACCESSKEY}" \
  EBS_SECRETKEY="${EBS_SECRETKEY}" \
  EBS_REGION="${EBS_REGION}"
```

**Comment by naviat** [comment]
> For anyone else having problems. Mine turned out to be me messing up creating the files by script.
> 
> So here are the scripts that I'm using that eventually worked on t3.xlarge2 (Ubuntu 18.04)
> 
> Thanks to the previous comments which helped me find the problem :)
> 
> ### Make sure nvme is installed
> ```
> apt install nvme-cli -y
> ```
> 
> ### Add/etc/udev/rules.d/999-aws-ebs-nvme.rules
> ```
> echo '# /etc/udev/rules.d/999-aws-ebs-nvme.rules
> # ebs nvme devices
> KERNEL=="nvme[0-9]*n[0-9]*", ENV{DEVTYPE}=="disk", ATTRS{model}=="Amazon Elastic Block Store", PROGRAM="/usr/local/bin/ebs-nvme-mapping /dev/%k", SYMLINK+="%c"' > /etc/udev/rules.d/999-aws-ebs-nvme.rules
> ```
> 
> ### Add /usr/local/bin/ebs-nvme-mapping
> ```
> echo '#!/bin/bash
> # /usr/local/bin/ebs-nvme-mapping
> vol=$(/usr/sbin/nvme id-ctrl --raw-binary "$1" | cut -c3073-3104 | tr -s "'" "'" | sed "'"s/ $//g"'")
> vol=${vol#/dev/}
> if [[ -n "$vol" ]]; then
>   echo ${vol/xvd/sd} ${vol/sd/xvd}
> fi' > /usr/local/bin/ebs-nvme-mapping
> chmod u+x /usr/local/bin/ebs-nvme-mapping
> ```
> 
> ### Make sure rules have run or reboot
> udevadm control --reload-rules && udevadm trigger
> 
> ### Using rexray/ebs:edge
> ```
> docker plugin install --grant-all-permissions rexray/ebs:edge \
>   EBS_ACCESSKEY="${EBS_ACCESSKEY}" \
>   EBS_SECRETKEY="${EBS_SECRETKEY}" \
>   EBS_REGION="${EBS_REGION}"
> ```

Sorry for requoting, following with your step but still got an error when running a container with rexray/ebs:edge 

After run `docker plugin install` 

```
ubuntu@ip-172-31-30-168:~/test$ docker run -v test-vol-1:/data busybox mount | grep "/data"
docker: Error response from daemon: error while mounting volume '/var/lib/docker/plugins/d21e3fcfd1027e807ab36dbc46f9fe36e2f95c578b242604c07343ec010b341a/rootfs': VolumeDriver.Mount: docker-legacy: Mount: test-vol-1: failed: no device name returned.
ERRO[0000] error waiting for container: context canceled
```

**Comment by timhaak** [comment]
@naviat Have you restarted the server after making the change.

Can you see the new linked dev devices in /dev/ from the udev rule above?

Also, look at your system logs.

**Comment by ozonni** [comment]
isn't it fixed in this PR https://github.com/rexray/rexray/pull/1252 ?

**Comment by Jackyen** [comment]
still see this error .

**Comment by JavierPandu** [comment]
Yes, it does not work for me either. I think this project has been abandoned though, so I would not put much faith on this being fixed

**Comment by ozonni** [comment]
actually this plugin worked for me. I managed to mount EBS directly to docker running in ECS cluster. Although i ended up switching to EFS to avoid single AZ's restriction. And mounting EFS is now supported by ECS out of the box.

**Comment by treksler** [comment]
> Yes, it does not work for me either. I think this project has been abandoned though, so I would not put much faith on this being fixed

@timhaak instructions work for me using ubuntu instances and rexray/ebs:latest

**Comment by tomalok** [comment]
I've had rexray ebs working for some time, but recently saw my dockerd.log fill up with the warnings.

Apparently the regexps [here on line 32 and 50](https://github.com/rexray/rexray/blob/master/libstorage/drivers/storage/ebs/utils/utils_unix.go#L39-L50) weren't updated, and since the device is is using its realpath of `/dev/nvme...` instead of the compatibility symlinks managed by udev/mdev, that's why we get the warning but everything's still working.

@akutz @clintkitson @codenrhoden - If someone submitted a PR to clean this up (shouldn't be difficult), what's the chances that we'd get a new release?

**Comment by ghost** [comment]
For people still struggling with it (like me), here is a great guy giving very good details: https://github.com/oogali/ebs-automatic-nvme-mapping.

Do not get amazed by the various examples. The solution is simple, very similar, and short.

What made the difference for me was (among other things) to set "**SUBSYSTEM=="block",** " before the KERNEL assignment in the .rules file.

Rexray docs did not work for me, but this did.

**Comment by darkl0rd** [comment]
Interesting little fact, nitro instances only support up to 26 NVMe volumes. Attempting to attach a 27th will result in the volume hanging in an "attaching" state on the AWS side. I found this out the hard way, as a result a large scale outage. I always relied on the fact that I could attach more volumes (UseLargeDeviceRange) on a given node.

It would be very convenient if the RexRay driver would inform the swarm that the respective node is no longer eligible for scheduling, since it met the resource limit (26 volumes).

**Comment by ghost** [comment]
@darkl0rd didn't try it yet, but [portworx](https://portworx.com/ebs-stuck-attaching-state-docker-containers/) seems to tackle this problem with a single ebs per node with virtual volumes inside
